src/main/scala/org/scanet/optimizers/SGD.scala

```diff
@@ -0,0 +1,26 @@
+package org.scanet.optimizers
+
+import org.scanet.core.{Output, Tensor, TensorType}
+import org.scanet.math.Numeric
+import org.scanet.math.syntax._
+
+case class SGD(rate: Double = 0.01, momentum: Double = 0.0, nesterov: Boolean = false) extends Algorithm[Tensor[Float]] {
+
+  override def delta[A: Numeric : TensorType, B: Numeric : TensorType](f: TensorFunction[A, B], initArg: Output[A], prevState: Tensor[Float]): (Output[Float], Tensor[Float]) = {
+    // Nesterov == true:
+    //   v[i] = momentum * v[i-1] + rate * grad(arg - momentum * v[i-1])
+    // Nesterov == false:
+    //   v[i] = momentum * v[i-1] + rate * grad(arg)
+    // momentum == 0.0 (vanilla SGD):
+    //   v[i] = rate * grad(arg)
+    // arg = arg - v[i]
+
+    val m = momentum.toFloat.const
+    // TODO: should this be + for maximize?
```

Comment:
Calculating Nesterow with formula from the paper is problematic and will not allow optimizations in the future. What if we take an implementation from Keras where (see https://arxiv.org/pdf/1212.0901v2.pdf):
```
for p, g, m in zip(params, grads, moments):
            v = self.momentum * m - lr * g  # velocity
            self.updates.append(K.update(m, v))

            if self.nesterov:
                new_p = p + self.momentum * v - lr * g
            else:
                new_p = p + v
```

Reply:
Also see https://keras.io/api/optimizers/sgd/#sgd-class